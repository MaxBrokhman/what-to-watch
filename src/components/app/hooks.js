import {useEffect} from 'react';

import {
  setActiveMovie,
  setError,
  setFilms,
} from '../../reducer/actions';
import {sendRequest} from '../../api/api';

export const useActiveMovie = (dispatch) => {
  const startVideoButtonHandler = (movie) => (evt) => {
    evt.preventDefault();
    setActiveMovie(movie, dispatch);
  };

  const exitButtonHandler = (evt) => {
    evt.preventDefault();
    setActiveMovie(null, dispatch);
  };

  return {
    startVideoButtonHandler,
    exitButtonHandler,
  };
};

export const useFetchedFilms = (dispatch) => {
  const onFailure = () => {
    setError({
      message: `Films are not available right now. Please, try again later`,
    }, dispatch);
  };
  useEffect(() => {
    sendRequest({
      url: `/films`,
      onSuccess: setFilms,
      onFailure,
      dispatch,
      method: `get`,
    });
  }, []);
};
