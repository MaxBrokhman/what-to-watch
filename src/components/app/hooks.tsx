import {useEffect, SyntheticEvent} from 'react';

import {
  setError,
  setFilms,
  setIsAuthorizationRequired,
} from '../../reducer/actions';
import {api, sendRequest} from '../../api/api';
import {useHistory} from 'react-router-dom';
import {TDispatch} from '../../reducer/types';

type TUseActiveMovie = {
  startVideoButtonHandler: (id: number) => (evt: SyntheticEvent) => void;
  exitButtonHandler: (evt: SyntheticEvent) => void;
}

export const useActiveMovie = (): TUseActiveMovie => {
  const history = useHistory();
  const startVideoButtonHandler = (id: number) => (evt: SyntheticEvent): void => {
    evt.preventDefault();
    history.push(`watch/${id}`);
  };

  const exitButtonHandler = (evt: SyntheticEvent): void => {
    evt.preventDefault();
    history.goBack();
  };

  return {
    startVideoButtonHandler,
    exitButtonHandler,
  };
};

export const useFetchedFilms = (dispatch: TDispatch): void => {
  const onFailure = (): void => {
    setError({
      message: `Films are not available right now. Please, try again later`,
    }, dispatch);
  };
  api.interceptors.response.use(null, (response) => {
    if (response.status === 401) {
      setIsAuthorizationRequired(true, dispatch);
    }
  });
  useEffect(() => {
    sendRequest({
      url: `/films`,
      onSuccess: setFilms,
      onFailure,
      dispatch,
      method: `get`,
    });
  }, []);
};
