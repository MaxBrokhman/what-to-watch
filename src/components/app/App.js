import React, {Fragment, useReducer} from 'react';

import {MainPage} from '../main-page/main-page';
import {MoviePlayer} from '../movie-player/movie-player';
import {MainMovieCard} from '../main-movie-card/main-movie-card';
import {useActiveMovie, useFetchedFilms} from './hooks';
import {
  reducer,
  initialState,
  Context
} from '../../reducer/reducer';

export const App = () => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const {startVideoButtonHandler, exitButtonHandler} = useActiveMovie(dispatch);
  useFetchedFilms(dispatch);
  return (
    <Context.Provider value={{state, dispatch}}>
      {
        state.activeMovie
          ? <MoviePlayer onVideoExit={exitButtonHandler} movie={state.activeMovie} />
          : (<Fragment>
            {
              state.filmsList.length && (
                <MainMovieCard
                  onVideoStart={startVideoButtonHandler}
                  movie={state.filmsList[0]}
                />
              )
            }
            <MainPage />
          </Fragment>
          )
      }
    </Context.Provider>
  );
};
