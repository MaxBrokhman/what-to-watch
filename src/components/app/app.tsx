import React, {useReducer} from 'react';
import {
  Switch,
  Route,
  Redirect,
} from 'react-router-dom';

import {MainPage} from '../main-page/main-page';
import {MoviePlayer} from '../movie-player/movie-player';
import {MovieDetails} from '../movie-details/movie-details';
import {useActiveMovie, useFetchedFilms} from './hooks';
import {
  reducer,
  initialState,
  Context
} from '../../reducer/reducer';
import {SignIn} from '../sign-in/sign-in';
import {AddReview} from '../add-review/add-review';
import {MyList} from '../my-list/my-list';

export const App = (): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const {startVideoButtonHandler, exitButtonHandler} = useActiveMovie();
  useFetchedFilms(dispatch);
  return (
    <Context.Provider value={{state, dispatch}}>
      <Switch>
        <Route path='/login'>
          <SignIn />
        </Route>
        {
          state.filmsList.length && <Route
            path='/movie-card/:id'
            render={({match}): JSX.Element => (
              <MovieDetails id={match.params.id} startVideoButtonHandler={startVideoButtonHandler} />
            )}
            exact
          />
        }
        {
          state.filmsList.length && <Route path='/movie-card/:id/add-review' render={({match}): JSX.Element => (
            <AddReview id={match.params.id} />
          )} />
        }
        {
          state.filmsList.length && <Route path='/watch/:id' render={({match}): JSX.Element => {
            return <MoviePlayer
              id={match.params.id}
              onVideoExit={exitButtonHandler}
            />;
          }}/>
        }
        <Route path='/my-list'>
          <MyList />
        </Route>
        <Route path='/*'>
          {
            state.isAuthorizationRequired && <Redirect exact from='/*' to='/login' />
          }
          {
            !state.activeMovie && !state.isAuthorizationRequired && (
              <MainPage startVideoButtonHandler={startVideoButtonHandler} />
            )
          }
        </Route>
      </Switch>
    </Context.Provider>
  );
};
