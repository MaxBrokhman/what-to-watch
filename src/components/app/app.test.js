import React from 'react';
import renderer from 'react-test-renderer';

import {App} from './app';

it(`Renders App component correctly`, () => {
  const tree = renderer
                .create(
                    <App />
                )
                .toJSON();
  expect(tree).toMatchSnapshot();
});
