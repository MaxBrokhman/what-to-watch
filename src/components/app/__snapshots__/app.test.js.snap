// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders App component correctly 1`] = `
Array [
  "0",
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            all genres
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      />
    </section>
  </div>,
]
`;
